### ISSUE TITLE
The result of coco pretrained model is very poor

### ISSUE THREAD

**Issue by Pluto1314** [ucas-vg/TinyBenchmark#24]
Hi, I'm sorry to disturb you. 
When I train to get the coco pretrained model, I get very poor results.
# Just like this:
![coco results](https://user-images.githubusercontent.com/47765770/82730301-2f417a00-9d31-11ea-9ec7-f779b12b5c31.JPG)

# my training command:
export NGPUS=4
CUDA_VISIBLE_DEVICES=4,5,6,7 python -m torch.distributed.launch --nproc_per_node=$NGPUS tools/train_test_net.py

# and my yaml file is:

MODEL:
  META_ARCHITECTURE: "GeneralizedRCNN"
  WEIGHT: "catalog://ImageNetPretrained/MSRA/R-50"
  BACKBONE:
    CONV_BODY: "R-50-FPN"
  RESNETS:
    BACKBONE_OUT_CHANNELS: 256
  RPN:
    USE_FPN: True
    ANCHOR_STRIDE: (4, 8, 16, 32, 64)
    ANCHOR_SIZES: (8.31, 12.5, 18.55, 30.23, 60.41)  # adjust here
    ASPECT_RATIOS: (0.5, 1.3, 2)                     # adjust here
    PRE_NMS_TOP_N_TRAIN: 2000
    PRE_NMS_TOP_N_TEST: 1000
    POST_NMS_TOP_N_TEST: 1000
    FPN_POST_NMS_TOP_N_TEST: 1000
  ROI_HEADS:
    USE_FPN: True
  ROI_BOX_HEAD:
    POOLER_RESOLUTION: 7
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    POOLER_SAMPLING_RATIO: 2
    FEATURE_EXTRACTOR: "FPN2MLPFeatureExtractor"
    PREDICTOR: "FPNPredictor"
INPUT:
  USE_SCALE: True
  SCALES: (0.25,)     # here only used for test dataset transform, when scale match on
DATASETS:
  TRAIN: ("coco_2014_train", "coco_2014_valminusminival", "coco_2014_minival")
  TEST: ("coco_2014_minival",)
DATALOADER:
  SIZE_DIVISIBILITY: 32
  USE_MORE_DA: 4
  USE_SCALE_MATCH: True
  SCALE_MATCH:
    TYPE: 'MonotonicityScaleMatch'
    SOURCE_ANNO_FILE: '/home/zhangyc/datasets/coco/annotations/instances_simple_merge2014.json'
    TARGET_ANNO_FILE: '/home/zhangyc/datasets/Tiny/train/annotations/train.json'
    BINS: 100
    DEFAULT_SCALE: 0.25
    SCALE_RANGE: (0.1, 1.)
    OUT_SCALE_DEAL: 'clip'

SOLVER:
  WARMUP_METHOD: "constant"
  TEST_ITER: 10000
  CHECKPOINT_PERIOD: 30000
  BASE_LR: 0.02
  WEIGHT_DECAY: 0.0001
  STEPS: (60000, 80000)
  MAX_ITER: 90000
  IMS_PER_BATCH: 16
  NUM_GPU: 4

TEST:
  IMS_PER_BATCH: 16

OUTPUT_DIR: 'outputs/tiny_coco/FPN_msm_tiny_b4'

# and my training loss is also high:
![loss1](https://user-images.githubusercontent.com/47765770/82730694-4a61b900-9d34-11ea-834f-8a4790482bd7.JPG)

![loss](https://user-images.githubusercontent.com/47765770/82730680-20a89200-9d34-11ea-88b0-5c345b609251.JPG)

my log.txt is: [log.txt](https://github.com/ucas-vg/TinyBenchmark/files/4671748/log.txt)

# question:

May I ask if my parameters or training methods are wrong.
Can you give me some suggestion? Thanks!
Look forward to your reply!

**Comment by yinglang** [comment]
Thank you for giving the very great question. I will try to give my understand.
1. If you can make sure there are no any technical mistake, your result probably be ok, cause my result is also low while mon-scale match COCO to TinyPerson, like
`2019-12-10 00:06:02,182 maskrcnn_benchmark.trainer INFO: eta: 0:00:56  iter: 44720  loss: 0.9321 (1.0735)  loss_retina_cls: 0.3358 (0.4309)  loss_retina_reg: 0.5869 (0.6426)  time: 0.1733 (0.2021)  data: 0.0189 (0.0520)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:05,477 maskrcnn_benchmark.trainer INFO: eta: 0:00:52  iter: 44740  loss: 0.8741 (1.0734)  loss_retina_cls: 0.3316 (0.4309)  loss_retina_reg: 0.5391 (0.6425)  time: 0.1657 (0.2021)  data: 0.0172 (0.0520)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:09,019 maskrcnn_benchmark.trainer INFO: eta: 0:00:48  iter: 44760  loss: 0.9083 (1.0733)  loss_retina_cls: 0.3495 (0.4308)  loss_retina_reg: 0.5624 (0.6425)  time: 0.1682 (0.2021)  data: 0.0175 (0.0520)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:12,378 maskrcnn_benchmark.trainer INFO: eta: 0:00:44  iter: 44780  loss: 0.8938 (1.0733)  loss_retina_cls: 0.3365 (0.4308)  loss_retina_reg: 0.5539 (0.6425)  time: 0.1639 (0.2021)  data: 0.0166 (0.0520)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:15,681 maskrcnn_benchmark.trainer INFO: eta: 0:00:40  iter: 44800  loss: 0.8911 (1.0732)  loss_retina_cls: 0.3332 (0.4307)  loss_retina_reg: 0.5683 (0.6424)  time: 0.1636 (0.2020)  data: 0.0177 (0.0519)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:18,965 maskrcnn_benchmark.trainer INFO: eta: 0:00:36  iter: 44820  loss: 0.9259 (1.0731)  loss_retina_cls: 0.3436 (0.4307)  loss_retina_reg: 0.5758 (0.6424)  time: 0.1645 (0.2020)  data: 0.0172 (0.0519)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:22,336 maskrcnn_benchmark.trainer INFO: eta: 0:00:32  iter: 44840  loss: 0.8854 (1.0730)  loss_retina_cls: 0.3422 (0.4307)  loss_retina_reg: 0.5447 (0.6424)  time: 0.1675 (0.2020)  data: 0.0115 (0.0519)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:25,647 maskrcnn_benchmark.trainer INFO: eta: 0:00:28  iter: 44860  loss: 0.9016 (1.0729)  loss_retina_cls: 0.3385 (0.4306)  loss_retina_reg: 0.5648 (0.6423)  time: 0.1636 (0.2020)  data: 0.0157 (0.0519)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:29,081 maskrcnn_benchmark.trainer INFO: eta: 0:00:24  iter: 44880  loss: 0.9047 (1.0729)  loss_retina_cls: 0.3378 (0.4306)  loss_retina_reg: 0.5629 (0.6423)  time: 0.1698 (0.2020)  data: 0.0164 (0.0519)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:32,453 maskrcnn_benchmark.trainer INFO: eta: 0:00:20  iter: 44900  loss: 0.8851 (1.0728)  loss_retina_cls: 0.3276 (0.4305)  loss_retina_reg: 0.5491 (0.6423)  time: 0.1624 (0.2020)  data: 0.0153 (0.0519)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:35,635 maskrcnn_benchmark.trainer INFO: eta: 0:00:16  iter: 44920  loss: 0.9114 (1.0727)  loss_retina_cls: 0.3411 (0.4305)  loss_retina_reg: 0.5610 (0.6422)  time: 0.1595 (0.2020)  data: 0.0184 (0.0518)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:38,949 maskrcnn_benchmark.trainer INFO: eta: 0:00:12  iter: 44940  loss: 0.8918 (1.0726)  loss_retina_cls: 0.3440 (0.4305)  loss_retina_reg: 0.5430 (0.6422)  time: 0.1581 (0.2019)  data: 0.0142 (0.0518)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:42,209 maskrcnn_benchmark.trainer INFO: eta: 0:00:08  iter: 44960  loss: 0.9136 (1.0726)  loss_retina_cls: 0.3438 (0.4304)  loss_retina_reg: 0.5636 (0.6421)  time: 0.1635 (0.2019)  data: 0.0154 (0.0518)  lr: 0.000200  max mem: 3172
2019-12-10 00:06:45,462 maskrcnn_benchmark.trainer INFO: eta: 0:00:04  iter: 44980  loss: 0.9182 (1.0725)  loss_retina_cls: 0.3419 (0.4304)  loss_retina_reg: 0.5729 (0.6421)  time: 0.1606 (0.2019)  data: 0.0164 (0.0518)  lr: 0.000200  max mem: 3172
....
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.1353
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.2665
 Average Precision  (AP) @[ IoU=0.60      | area=   all | maxDets=100 ] = 0.2160
 Average Precision  (AP) @[ IoU=0.70      | area=   all | maxDets=100 ] = 0.1561
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.1223
 Average Precision  (AP) @[ IoU=0.80      | area=   all | maxDets=100 ] = 0.0884
 Average Precision  (AP) @[ IoU=0.90      | area=   all | maxDets=100 ] = 0.0185
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=smallest | maxDets=100 ] = 0.0365
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=  fpn1 | maxDets=100 ] = 0.1097
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=  fpn2 | maxDets=100 ] = 0.1657
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=  fpn3 | maxDets=100 ] = 0.2427
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=fpn4+5 | maxDets=100 ] = 0.3061
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=largest | maxDets=100 ] = -1.0000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.1662
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.2555
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.2690
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=smallest | maxDets=100 ] = 0.0784
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=  fpn1 | maxDets=100 ] = 0.2613
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=  fpn2 | maxDets=100 ] = 0.3428
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=  fpn3 | maxDets=100 ] = 0.4435
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=fpn4+5 | maxDets=100 ] = 0.6048
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=largest | maxDets=100 ] = -1.0000
2019-12-10 00:09:51,996 maskrcnn_benchmark.inference INFO: OrderedDict([('bbox', OrderedDict([('AP', 0.13526422595582285), ('AP50', 0.2665151996355452), ('AP75', 0.21595325953126962), ('APs', 0.15608672323490513), ('APm', 0.12231351193679162), ('APl', 0.08839402837425384)]))])`
2. why AP on COCO test is low? Cause the input test image is low resolution which set in config file(the input image size is in (100, 167) in this setting while size of normal coco is in (800, 1333)).
`
INPUT:
USE_SCALE: True
SCALES: (0.25,) # here only used for test dataset transform, when scale match on
`
3. Is AP on COCO test set important? No, the target of MSM pretrain is not trying to improve performance on COCO test set, but make a better pretrain for your dataset. So in config file, it use distribute of your dataset. low performance on COCO test does not mean low performance on your dataset, you need try.
`
SCALE_MATCH:
TYPE: 'MonotonicityScaleMatch'
SOURCE_ANNO_FILE: '/home/zhangyc/datasets/coco/annotations/instances_simple_merge2014.json'
TARGET_ANNO_FILE: '/home/zhangyc/datasets/Tiny/train/annotations/train.json'
`
4. How can I use AP on COCO test set? I think AP on COCO test can use to debug. Cause if train COCO train set with size in (100, 167), you will get about AP=14,AP50=30(not fine adjust), So if the size range of your dataset is closer to (100, 167), you should get closer performance to this, otherwise lower performance.

**Comment by Pluto1314** [comment]
Thank you very much for your detailed reply!
I will try to train with my own dataset. If I have some questions, I may continue to disturb you.

**Comment by Pluto1314** [comment]
@yinglang Hi, I want to trian on my onw datasets with the pretrained MSM COCO weights.
Which yaml file should I choose as the config file? 
Should I choose the e2e_faster_rcnn_R_50_FPN_1x_cocostyle_msmb4coco.yaml or other yaml file?
Because when I was training, the program started testing directly.
# just like this:
![1](https://user-images.githubusercontent.com/47765770/82779582-2498f900-9e87-11ea-9caf-fd436484e70b.JPG)

# my yaml file is:
MODEL:
  META_ARCHITECTURE: "GeneralizedRCNN"
  WEIGHT: '/home/zhangyc/TinyBenchmark/tiny_benchmark/outputs/tiny_coco/FPN_msm_tiny_b4/model_final.pth'    # MSM COCO pretrained weights
  IGNORE_WEIGHT_KEYS: ["roi_heads.box.predictor"]
  BACKBONE:
    CONV_BODY: "R-50-FPN"

  RESNETS:
    BACKBONE_OUT_CHANNELS: 256
  RPN:
    USE_FPN: True
    ANCHOR_STRIDE: (4, 8, 16, 32, 64)
    PRE_NMS_TOP_N_TRAIN: 2000
    PRE_NMS_TOP_N_TEST: 1000
    POST_NMS_TOP_N_TEST: 1000
    FPN_POST_NMS_TOP_N_TEST: 1000
    ANCHOR_SIZES: (8.31, 12.5, 18.55, 30.23, 60.41)  # (4, 8, 16, 32, 64) will be better
    ASPECT_RATIOS: (0.5, 1.3, 2)
  ROI_HEADS:
    USE_FPN: True
    DETECTIONS_PER_IMG: 200
  ROI_BOX_HEAD:
    POOLER_RESOLUTION: 7
    POOLER_SCALES: (0.25, 0.125, 0.0625, 0.03125)
    POOLER_SAMPLING_RATIO: 2
    FEATURE_EXTRACTOR: "FPN2MLPFeatureExtractor"
    PREDICTOR: "FPNPredictor"
    NUM_CLASSES: 2                                                    # not forget
INPUT:
  USE_SCALE: true
  SCALES: (1,)
DATASETS:
  TRAIN: ("tiny_train_coco",)
  TEST: ("tiny_test_coco",)
DATALOADER:
  SIZE_DIVISIBILITY: 32

SOLVER:
  CHECKPOINT_PERIOD: 40000   
  TEST_ITER: 10000   
  BASE_LR: 0.02
  WEIGHT_DECAY: 0.0001
  STEPS: (40000, 60000)   
  MAX_ITER: 80000 
  IMS_PER_BATCH: 2
  NUM_GPU: 2
TEST:
  IMS_PER_BATCH: 2
  COCO_EVALUATE_STANDARD: 'tiny'  # tiny need change
  MERGE_RESULTS: false
  MERGE_GT_FILE: ""
  IGNORE_UNCERTAIN: false
  USE_IOD_FOR_IGNORE: false
OUTPUT_DIR: ./outputs/tiny_set/FPN/baseline2_msmb4coco

**Comment by Pluto1314** [comment]
I will encounter the above-mentioned problems when I use the MSM COCO model I trained myself, but the above-mentioned problems will not happen if I use the pre-trained model you provide.

In the end I got a okay result.

![ap](https://user-images.githubusercontent.com/47765770/82862147-845bd680-9f51-11ea-866e-ec1bca7d09bd.JPG)

![1454_03](https://user-images.githubusercontent.com/47765770/82862174-95a4e300-9f51-11ea-875a-f0cddbc4b11f.jpg)

Thank you very much for your reply every time, it has given me great help!

**Comment by yinglang** [comment]
We are glad that our work can help you.  For your problem, adjust MAX_ITER is not a good idea. It happened because of your saved model have other checkpoint information(like optimizer), you need delete them first, just like following code showing. And 'model_only_model_final.pth' will be ok.

`py
ckpt_data = torch.load('model_final.pth', map_location=torch.device("cpu"))
[ckpt_data.pop(k) for k in kept_data if k != 'model']
torch.save(checkpointer_data,'model_only_model_final.pth')
`

For the result of your dataset, it looks great. But there are still some miss detections(False negative). And there are too much target in a single image, if you have not cropped it to pieces while testing, You may need adjust these param in config file while training:

> PRE_NMS_TOP_N_TRAIN: 2000

and adjust there param while testing
>RPN:
  PRE_NMS_TOP_N_TEST: 1000
  POST_NMS_TOP_N_TEST: 1000
  FPN_POST_NMS_TOP_N_TEST: 1000
ROI_HEADS:
   SCORE_THRESH = 0.05
   DETECTIONS_PER_IMG = 100

**Comment by Pluto1314** [comment]
Thank you for your reply, I will try it.
At the same time, I plan to use retinanet to implement the effect. Because I found that faster rcnn is slower, retinanet should be faster.

**Comment by abhinavchawla13** [comment]
@Pluto1314, can you help me with 2 questions:
1. How were you able to test it against your own dataset
2. How to add boxes onto an image after getting predictions with bbox
I'm new to it, so am a bit confused. Any help would be appreciated!

**Comment by Pluto1314** [comment]
@abhinavchawla13 I'm sorry to reply you so late.
1、I train with my own dataset.
2、may be you can try https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/demo/webcam.py

**Comment by 422301949** [comment]
@Pluto1314 Hello！Your test result looks very good! I am very interested about your work.  Are you directly using the pre-trained model by the author and then fine-tuing on your dataset without any tick?

**Comment by Pluto1314** [comment]
@422301949 Yes, I used the COCO pre-trained model trained with my own dataset, and then used this model to fine-tuing the dataset. And, I used the xView.
